src/psyclone/psyGen.py

```diff
@@ -2442,29 +2459,47 @@ def call(self, value):
     def backward_dependence(self):
         '''Returns the preceding argument that this argument has a direct
         dependence with, or None if there is not one. The argument may
-        exist in a call, a haloexchange, or a globalsum. For performance
-        reasons only compute once then store the result. '''
-        if not self._bd_computed:
-            self._bd_computed = True
-            all_nodes = self._call.walk(self._call.root.children, Node)
-            position = all_nodes.index(self._call)
-            all_prev_nodes = all_nodes[:position]
-            all_prev_nodes.reverse()
-            self._bd_value = self._find_argument(all_prev_nodes)
-        return self._bd_value
+        exist in a call, a haloexchange, or a globalsum. '''
+        all_nodes = self._call.walk(self._call.root.children, Node)
+        position = all_nodes.index(self._call)
+        all_prev_nodes = all_nodes[:position]
+        all_prev_nodes.reverse()
+        return self._find_argument(all_prev_nodes)
+
+    def backward_write_dependencies(self, ignore_halos=False):
```

Comment:
These have different functionality. backward_dependence will find the nearest previous dependence for a field with any access. So if it is a read it will find the nearest write and if it is a write it will find the nearest read. backward_write_dependencies returns all previous write dependencies for a field and will return an empty list if the the field writes. I can gather the 4 common statements to both methods into a function if you think that would be better.